Add Target Functions (#126)
CMakePP/CMaize: docs/src/developer/design/user_api/add_target.rst

functions will necessarily wrap CMake's add target functions.

The primary reason CMaize needs wrapper functions is to capture the user's
input. More specifically, the CMaize versions of the add target functions will
record the target options in the active ``CMaizeProject`` so that CMaize will
be able to automate packaging of the project. Another motivation for
maintaining a separate set of CMaize add target functions is to
provide more succinct APIs. In practice, declaring CMake targets can be very
verbose and much of what needs to be provided can be inferred or learned from
other sources.

**********************
Add Target Terminology
**********************

CMaize will be dealing with targets associated with multiple coding languages.
Most coding languages organize code similarly, but differ in what they term
those organizational units. In order to provide a unified description we define
the following terms. We use these terms throughout this page regardless of what
the coding-language appropriate terms are.

.. glossary::

   executable
      A program meant to be run by a user. It may be compiled or it may not be,

>       e.g., a Python script with a ``main`` function.

Reviewer: ```suggestion
      e.g., a Python script.
```

Technically the Python script does not need a `main` function to be run through something like `python <script_name>.py`. Although I think a `main()` function is useful, you could just write a few lines of procedural code in a file and then execute the file, right?

   library
      A collection of functionality distributed as a single packaged entity.
      Libraries are designed for consumption by other packages.

   source file
      Any file containing code. For languages like C++, "source file" includes
      header files.

As a slight aside, we choose these terms in order to conform to CMake's already
existing :term:`API`.

**********************************
Add Target Function Considerations
**********************************

.. _at_recording_targets:

recording targets
   The primary motivation for the add target functions is to serve as a
   mechanism for recording the details of the target.

.. _at_succinctness:

succinctness

>    A lot of the information CMake's add target functions require can be

Reviewer: ```suggestion
   A lot of the information required by CMake's add target functions can be
```
For clarity.

   gleaned from other sources. Requiring the user to restate the information is
   verbose and violates :term:`DRY`.

.. _at_coding_languages:

coding language
   Exactly what targets can be built/found depend on the coding language(s) of
   the project. For example, shared/static :term:`libraries <library>` do not

>    exist in the context of building a Python code, but do exist when building

Reviewer: ```suggestion
   exist in the context of building Python code, but do exist when building
```

     target, whereas private source files are consumed in building the target.

.. _at_conditional_targets:

conditional targets
   Many projects contain targets which are only conditionally built. These
   targets may be optional package features, or targets only needed for testing
   or maintaining the project. In our experience, in a traditional CMake-based
   :term:`build system`, it is rarely possible to isolate the logic for these
   targets because CMake requires them to be specified in multiple places. We
   assume the following about a conditional target:

   - is conditionally included based on the value of a variable (if it's
     actually multiple variables, the user, via boolean logic, can combine the
     variables into a single variable). For optional features the variable is
     usually something like ``ENABLE_XXX``; for tests the variable is
     ``BUILD_TESTING`` (defined by CMake).
   - needs to be conditionally built, linked to, tested, and packaged. In other
     words, when CMaize is given a list of targets, CMaize needs to skip
     conditional targets which are not currently enabled.

> 

Reviewer: I would suggest adding another consideration here, something like "maintaining best practices" saying:

Along with the concerns addressed in consideration `succinctness`, creating a new target in CMake has different best practices for different languages and different target systems that often must be inferred or learned from other sources.

Author: I like this idea, though I'm not sold on your wording (or the wording I just pushed). Let me know what you think.

Reviewer: I think your new wording is fine!

In this sense, CMaize's add target functions can be thought of as factory
routines.

***************
Add Target APIs
***************

The APIs for ``cmaize_add_executable`` and ``cmaize_add_library`` were
briefly introduced in the :ref:`designing_cmaizes_user_api` section. Generally
speaking the two functions work similarly aside from the fact that
``cmaize_add_executable``/\ ``cmaize_add_library`` will ultimately map to
CMake's ``add_executable``/\ ``add_library`` function, respectively (the calls
to the CMake functions happen inside the target objects). Given the similar APIs
we presently limit our focus to ``cmaize_add_library``:

.. code-block:: CMake

   # A typical C++ invocation
   cmaize_add_library(
       "<target name>"
       SOURCE_DIR "<directory containing private source files>"
       INCLUDE_DIRS "<directory(s) containing public source files>"
       DEPENDS "<name of dependency 0>" "<name of dependency 1>"
   )

   # A typical Python invocation
   cmaize_add_library(
       "<target name>"
       INCLUDE_DIRS "<directory(s) containing the Python library>"
       DEPENDS "<name of dependency 0>" "<name of dependency 1>"

>    )

Author: @zachcran what do you think of this API? I'm wondering if `INCLUDE_DIRS` for Python should map to `PYTHONPATH` and `SOURCE_DIR` should be the contents of the library?

Reviewer: I like `SOURCE_DIR` pointing to the library code. I'm not so sure about `INCLUDE_DIRS` mapping to `PYTHONPATH`. It could definitely work, but I think you usually want to set up a virtual environment and install everything there. It seems like making it this accessible would encourage users away from the best practices of venvs and installable packages. Maybe it could be pushed to an optional `PYTHONPATH` parameter?

Author: We can punt on `PYTHONPATH` for now. When managing hybrid language projects its not so easy as "install everything in a virtual environment", e.g., if I'm trying to test the Python bindings for the C++ library I just built it's going to be sitting in the build directory, not installed in the virtual environment.

Reviewer: For testing things like the Python bindings you mentioned, could we add something like `PYTHONPATH=$PYTHONPATH:/path/to/build/directory <actual python command>` when calling `add_test()` when CMaize creates Python tests.
